# Adversarial Search using Dominoes game

## Imports

In [1]:
from game.game_state import GameState
from agents.random_agent import RandomAgent
from agents.heuristic_agent import HeuristicAgent
from agents.smart_heuristic_agent import SmartHeuristicAgent
from agents.monte_carlo_agent import MonteCarloAgent

## Game Setup

In [2]:
def play_ai_game(agent0, agent1, agent2, agent3):
    game = GameState()
    game.setup()

    agents = [agent0, agent1, agent2, agent3]

    while not game.game_over:
        player = game.current_player
        agent = agents[player]

        # handle forced first move
        if game.board.is_empty() and game.forced_first_tile is not None:
            tile = game.forced_first_tile
            side = None
        else:
            tile, side = agent.choose_move(game)

        if tile is None:
            game.apply_pass()
        else:
            game.apply_move(tile, side)

    return game.winner, game.scores

## Simple test

In [3]:
winner, scores = play_ai_game(
    RandomAgent(),
    HeuristicAgent(),
    SmartHeuristicAgent(),
    MonteCarloAgent()
)

print("Winner:", winner)
print("Scores:", scores)

Winner: 1
Scores: {0: 0, 1: 28}


## Multiple game sessions 

In [4]:
def run_experiment(agentA, agentB, games=100):
    wins = {0: 0, 1: 0, None: 0}

    for _ in range(games):
        winner, _ = play_ai_game(
            agentA(), agentB(), agentA(), agentB()
        )
        wins[winner] += 1

    return wins

def win_rate(results):
    total = sum(results.values())
    return {k: round(v / total, 3) for k, v in results.items()}

## Experiment Comparisons 

In [5]:
print("Random vs Random")
res = run_experiment(RandomAgent, RandomAgent, 200)
print(res)
print("Win rate:", win_rate(res))

print("\nHeuristic vs Random")
res = run_experiment(HeuristicAgent, RandomAgent, 200)
print(res)
print("Win rate:", win_rate(res))

print("\nSmart Heuristic vs Random")
res = run_experiment(SmartHeuristicAgent, RandomAgent, 200)
print(res)
print("Win rate:", win_rate(res))

Random vs Random
{0: 92, 1: 106, None: 2}
Win rate: {0: 0.46, 1: 0.53, None: 0.01}

Heuristic vs Random
{0: 94, 1: 104, None: 2}
Win rate: {0: 0.47, 1: 0.52, None: 0.01}

Smart Heuristic vs Random
{0: 116, 1: 81, None: 3}
Win rate: {0: 0.58, 1: 0.405, None: 0.015}


## Monte Carlo Evaluation

In [6]:
print("\nMonte Carlo vs Random")
res = run_experiment(MonteCarloAgent, RandomAgent, 200)
print(res)
print("Win rate:", win_rate(res))

print("\nMonte Carlo vs Heuristic")
res = run_experiment(MonteCarloAgent, HeuristicAgent, 200)
print(res)
print("Win rate:", win_rate(res))

print("\nMonte Carlo vs Smart Heuristic")
res = run_experiment(MonteCarloAgent, SmartHeuristicAgent, 200)
print(res)
print("Win rate:", win_rate(res))


Monte Carlo vs Random
{0: 164, 1: 34, None: 2}
Win rate: {0: 0.82, 1: 0.17, None: 0.01}

Monte Carlo vs Heuristic
{0: 176, 1: 24, None: 0}
Win rate: {0: 0.88, 1: 0.12, None: 0.0}

Monte Carlo vs Smart Heuristic
{0: 164, 1: 35, None: 1}
Win rate: {0: 0.82, 1: 0.175, None: 0.005}


## Final Strong Comparison

In [7]:
GAMES = 500  
print(f"\nFINAL TEST ({GAMES} games)\n")

print("Monte Carlo vs Random")
res = run_experiment(MonteCarloAgent, RandomAgent, GAMES)
print(res)
print("Win rate:", win_rate(res))

print("\nMonte Carlo vs Heuristic")
res = run_experiment(MonteCarloAgent, HeuristicAgent, GAMES)
print(res)
print("Win rate:", win_rate(res))

print("\nMonte Carlo vs Smart Heuristic")
res = run_experiment(MonteCarloAgent, SmartHeuristicAgent, GAMES)
print(res)
print("Win rate:", win_rate(res))


FINAL TEST (500 games)

Monte Carlo vs Random
{0: 421, 1: 77, None: 2}
Win rate: {0: 0.842, 1: 0.154, None: 0.004}

Monte Carlo vs Heuristic
{0: 418, 1: 80, None: 2}
Win rate: {0: 0.836, 1: 0.16, None: 0.004}

Monte Carlo vs Smart Heuristic
{0: 425, 1: 72, None: 3}
Win rate: {0: 0.85, 1: 0.144, None: 0.006}


## Data Analysis 

    The experimental results show that the Monte Carlo agent consistently outperformed all other agents tested. Against the Random, Heuristic, and Smart Heuristic agents, the Monte Carlo approach achieved win rates above 83%, demonstrating strong and stable performance across different opponent strategies. The baseline experiments also showed that the standard Heuristic agent performed similarly to the Random agent, while the Smart Heuristic agent achieved better results with a 58% win rate against Random play. However, neither heuristic approach was able to compete effectively against the Monte Carlo agent, indicating that simulation-based decision-making was significantly more effective than fixed rule-based strategies.

## Conclusion 

    Overall, the project successfully demonstrated the effectiveness of the Monte Carlo simulation approach in adversarial AI gameplay. The Monte Carlo agent consistently achieved the highest win rates and proved to be both robust and adaptable against different opponents. The experiments also highlighted the importance of running a large number of games to obtain reliable and statistically meaningful results. Future improvements could include increasing the number of simulations, combining heuristics with Monte Carlo search, or exploring reinforcement learning techniques to further improve performance.